packages/client/src/components/SafeContacts.js

```diff
@@ -1,43 +1,191 @@
+      contact={{ name: '', address: '' }} 
+    />
+  );
+
+  const openEditModal = (index, contact) => modalContext.openModal(
+    <EditContactModal 
+      headerTitle="Edit Contact"
+      confirmText="Update"
+      onConfirm={(newContact) => setContact(index, newContact)}
+      contact={contact} 
+    />
+  );
+
+  const openRemoveModal = (index, contact) => modalContext.openModal(
+    <RemoveContactModal 
+      headerTitle="Confirm"
+      confirmText="Confirm"
+      onConfirm={() => removeContact(index)}
+      contact={contact} 
+    />
+  );
+
   return (
     <>
-      <div className="column p-0 mt-5 is-flex is-full">
-        <h4 className="is-size-5">Saved addresses</h4>
+      <div className="column p-0 mt-5 is-flex is-align-items-center is-justify-content-space-between is-full">
+        <h4 className="is-size-5">Saved Addresses</h4>
+        {!isEmpty(contacts) &&
+          <button className="button is-link" onClick={openAddModal}>
+            Add Contact <Plus className="ml-2" />
+          </button>
+        }
       </div>
-      <div className="column p-0 mt-4 is-flex is-flex-direction-column is-full rounded-sm border-light has-shadow">
-        {verifiedSafeOwners.map((so, i) => (
-          <div
-            className="is-flex is-align-items-center is-justify-content-space-between column is-full p-5 border-light-bottom"
-            key={i}
-          >
-            <div className="mr-5">
-              <Person />
+      {isEmpty(contacts) ? <EmptyContacts setContact={setContact} openAddModal={openAddModal} /> : (
+        <div className="column p-0 mt-4 is-flex is-flex-direction-column is-full rounded-sm border-light has-shadow">
+          <div className="is-flex is-align-items-center is-justify-content-space-between column is-full p-5 border-light-bottom">
+            <div className="mr-5" style={{ width: addressWidth }}>
+              Address
             </div>
-            <div className="px-2 mr-6" style={{ minWidth: 120 }}>
-              {so.name ?? `Signer #${i + 1}`}
+            <div className="px-2 mr-6" style={{ width: minWidth }}>
+              Name
             </div>
-            <div className="flex-2">{so.address}</div>
-            <div className="is-underlined">
-              <span className="mr-5">Edit</span>
-              <span
-                className="mr-5 pointer"
-                onClick={() => clipboard.copy(so.address)}
-              >
-                {clipboard.textJustCopied === so.address
-                  ? "Copied"
-                  : "Copy Address"}
-              </span>
-              <span>Send</span>
+            <div style={{ width: minWidth }}>
+              Actions
             </div>
           </div>
-        ))}
-      </div>
+          {contacts.map((contact, index) => (
+            <div
+              className="is-flex is-align-items-center is-justify-content-space-between column is-full p-5 border-light-bottom"
+              key={index}
```

Comment:
We should avoid using index as key, refer to the article here :https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#keys-and-reconciliation